internal/scheduler/reports.go

```diff
@@ -330,42 +287,147 @@ func (repo *SchedulingContextRepository) GetSchedulingReport(_ context.Context,
 	return &schedulerobjects.SchedulingReport{Report: report}, nil
 }
 
-type schedulingReport struct {
-	mostRecentByExecutor           SchedulingContextByExecutor
-	mostRecentSuccessfulByExecutor SchedulingContextByExecutor
-	mostRecentPreemptingByExecutor SchedulingContextByExecutor
-
-	sortedExecutorIds []string
-}
-
-func (sr schedulingReport) ReportString(verbosity int32) string {
+func (repo *SchedulingContextRepository) getSchedulingReportString(verbosity int32) string {
+	mostRecentByExecutor := repo.GetMostRecentSchedulingContextByExecutor()
+	mostRecentSuccessfulByExecutor := repo.GetMostRecentSuccessfulSchedulingContextByExecutor()
+	mostRecentPreemptingByExecutor := repo.GetMostRecentPreemptingSchedulingContextByExecutor()
 	var sb strings.Builder
 	w := tabwriter.NewWriter(&sb, 1, 1, 1, ' ', 0)
-	for _, executorId := range sr.sortedExecutorIds {
+	for _, executorId := range repo.GetSortedExecutorIds() {
 		fmt.Fprintf(w, "%s:\n", executorId)
-		if sctx := sr.mostRecentByExecutor[executorId]; sctx != nil {
-			fmt.Fprint(w, indent.String("\t", "Most recent attempt:\n"))
+		if sctx := mostRecentByExecutor[executorId]; sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent scheduling round:\n"))
 			fmt.Fprint(w, indent.String("\t\t", sctx.ReportString(verbosity)))
 		} else {
-			fmt.Fprint(w, indent.String("\t", "Most recent attempt: none\n"))
+			fmt.Fprint(w, indent.String("\t", "Most recent scheduling round: none\n"))
 		}
-		if sctx := sr.mostRecentSuccessfulByExecutor[executorId]; sctx != nil {
-			fmt.Fprint(w, indent.String("\t", "Most recent successful attempt:\n"))
+		if sctx := mostRecentSuccessfulByExecutor[executorId]; sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent scheduling round that scheduled a job:\n"))
 			fmt.Fprint(w, indent.String("\t\t", sctx.ReportString(verbosity)))
 		} else {
-			fmt.Fprint(w, indent.String("\t", "Most recent successful attempt: none\n"))
+			fmt.Fprint(w, indent.String("\t", "Most recent scheduling round that scheduled a job: none\n"))
 		}
-		if sctx := sr.mostRecentPreemptingByExecutor[executorId]; sctx != nil {
-			fmt.Fprint(w, indent.String("\t", "Most recent preempting attempt:\n"))
+		if sctx := mostRecentPreemptingByExecutor[executorId]; sctx != nil {
+			fmt.Fprint(w, indent.String("\t", "Most recent scheduling round that preempted a job:\n"))
 			fmt.Fprint(w, indent.String("\t\t", sctx.ReportString(verbosity)))
 		} else {
-			fmt.Fprint(w, indent.String("\t", "Most recent preempting attempt: none\n"))
+			fmt.Fprint(w, indent.String("\t", "Most recent scheduling round that preempted a job: none\n"))
+		}
+	}
+	w.Flush()
+	return sb.String()
```

Comment:
I've essentially stamped out a couple of copies of this code with minor variations, because I couldn't think of satisfying abstraction.